File: internal/browser.go
package internal

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/google/uuid"
)

type browserRequest struct {
	ID          string `json:"id"`
	Action      string `json:"action"`
	URL         string `json:"url,omitempty"`
	Ref         string `json:"ref,omitempty"`
	Text        string `json:"text,omitempty"`
	Attribute   string `json:"attribute,omitempty"`
	Script      string `json:"script,omitempty"`
	Value       string `json:"value,omitempty"`
	Interactive bool   `json:"interactive,omitempty"`
	Selector    string `json:"selector,omitempty"`
	TabID       any    `json:"tabId,omitempty"`
	Index       *int   `json:"index,omitempty"`
	Key         string `json:"key,omitempty"`
	Direction   string `json:"direction,omitempty"`
	Pixels      *int   `json:"pixels,omitempty"`
	Filter      string `json:"filter,omitempty"`
}

type browserResponseData struct {
	Title        string `json:"title,omitempty"`
	URL          string `json:"url,omitempty"`
	TabID        int    `json:"tabId,omitempty"`
	SnapshotData *struct {
		Snapshot string `json:"snapshot"`
	} `json:"snapshotData,omitempty"`
	Value          string `json:"value,omitempty"`
	ScreenshotPath string `json:"screenshotPath,omitempty"`
	DataURL        string `json:"dataUrl,omitempty"`
	Result         any    `json:"result,omitempty"`
	Tabs           []struct {
		Index  int    `json:"index"`
		URL    string `json:"url"`
		Title  string `json:"title"`
		Active bool   `json:"active"`
		TabID  int    `json:"tabId"`
	} `json:"tabs,omitempty"`
}

type browserResponse struct {
	ID      string               `json:"id"`
	Success bool                 `json:"success"`
	Data    *browserResponseData `json:"data,omitempty"`
	Error   string               `json:"error,omitempty"`
}

// InvokeBrowser sends a command to bb-browser daemon and returns the result.
func InvokeBrowser(endpoint string, req *browserRequest) (*browserResponse, error) {
	if req.ID == "" {
		req.ID = uuid.NewString()
	}

	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("marshal browser request: %w", err)
	}

	client := &http.Client{Timeout: 35 * time.Second}
	resp, err := client.Post(endpoint+"/command", "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("browser daemon unreachable: %w", err)
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode == 503 {
		return nil, fmt.Errorf("browser extension not connected")
	}
	if resp.StatusCode == 408 {
		return nil, fmt.Errorf("browser command timed out")
	}

	var result browserResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("parse browser response: %w", err)
	}
	return &result, nil
}

// saveScreenshotFromDataURL decodes a data:image/png;base64,... URL and saves to data/images/.
func saveScreenshotFromDataURL(dataURL string) (bool, string) {
	// Parse data URL: data:image/png;base64,<data>
	const prefix = "base64,"
	idx := strings.Index(dataURL, prefix)
	if idx < 0 {
		return false, ""
	}
	b64Data := dataURL[idx+len(prefix):]

	data, err := base64.StdEncoding.DecodeString(b64Data)
	if err != nil {
		return false, ""
	}

	// Determine extension from mime type
	ext := ".png"
	if strings.Contains(dataURL[:idx], "jpeg") {
		ext = ".jpg"
	} else if strings.Contains(dataURL[:idx], "webp") {
		ext = ".webp"
	}

	// Generate filename with timestamp
	filename := fmt.Sprintf("screenshot-%d%s", time.Now().UnixMilli(), ext)
	relPath := filepath.Join("images", filename)

	absPath, err := resolvePath(relPath)
	if err != nil {
		return false, ""
	}
	os.MkdirAll(filepath.Dir(absPath), 0o755)

	if err := os.WriteFile(absPath, data, 0o644); err != nil {
		return false, ""
	}

	url := pinixDataURLPrefix + relPath
	return true, url
}

// formatBrowserResult converts a browser response to a human/LLM readable string.
func formatBrowserResult(resp *browserResponse) string {
	if !resp.Success {
		return fmt.Sprintf("[error] %s", resp.Error)
	}
	if resp.Data == nil {
		return "OK"
	}

	var parts []string

	if resp.Data.SnapshotData != nil {
		return resp.Data.SnapshotData.Snapshot
	}

	if resp.Data.Title != "" {
		parts = append(parts, fmt.Sprintf("Title: %s", resp.Data.Title))
	}
	if resp.Data.URL != "" {
		parts = append(parts, fmt.Sprintf("URL: %s", resp.Data.URL))
	}
	if resp.Data.Value != "" {
		parts = append(parts, resp.Data.Value)
	}
	if resp.Data.ScreenshotPath != "" {
		parts = append(parts, fmt.Sprintf("Screenshot: %s", resp.Data.ScreenshotPath))
	}
	// Auto-save screenshot from dataUrl to data/images/
	if resp.Data.DataURL != "" {
		if saved, url := saveScreenshotFromDataURL(resp.Data.DataURL); saved {
			parts = append(parts, fmt.Sprintf("Render: ![screenshot](%s)", url))
		}
	}
	if resp.Data.Result != nil {
		b, _ := json.Marshal(resp.Data.Result)
		parts = append(parts, string(b))
	}

	if len(resp.Data.Tabs) > 0 {
		var b strings.Builder
		for _, tab := range resp.Data.Tabs {
			marker := " "
			if tab.Active {
				marker = "*"
			}
			fmt.Fprintf(&b, "%s [%d] %s — %s\n", marker, tab.TabID, tab.Title, tab.URL)
		}
		parts = append(parts, b.String())
	}

	if len(parts) == 0 {
		return "OK"
	}
	return strings.Join(parts, "\n")
}

// RegisterBrowserCommands adds browser commands to the registry.
func RegisterBrowserCommands(r *Registry, cfg *Config) {
	if cfg.Browser == nil || cfg.Browser.Endpoint == "" {
		return
	}
	endpoint := cfg.Browser.Endpoint

	r.Register("browser", `Control the user's Chrome browser via bb-browser.
  browser open <url>                  — open URL in current tab
  browser snapshot [--interactive]    — get page accessibility tree
  browser click <ref>                 — click element (e.g. @5 or 5)
  browser fill <ref> <text>           — clear input and fill text
  browser type <ref> <text>           — type text without clearing
  browser press <key>                 — press key (Enter, Tab, Escape, etc.)
  browser scroll <dir> [pixels]       — scroll: up/down/left/right (default 300)
  browser eval <script>               — execute JavaScript
  browser get text|url|title [<ref>]  — get element text or page info
  browser screenshot                  — take screenshot
  browser close                       — close current tab
  browser back|forward|refresh        — navigate
  browser tabs                        — list open tabs
  browser tab-new [url]               — open new tab
  browser tab-select <tabId>          — switch to tab
  browser tab-close [tabId]           — close tab`,
		func(args []string, stdin string) (string, error) {
			if len(args) == 0 {
				return "", fmt.Errorf("usage: browser <action> [args...]")
			}

			req := parseBrowserArgs(args)
			if req == nil {
				return "", fmt.Errorf("unknown browser action: %s", args[0])
			}

			resp, err := InvokeBrowser(endpoint, req)
			if err != nil {
				return "", err
			}
			return formatBrowserResult(resp), nil
		})
}

func parseBrowserArgs(args []string) *browserRequest {
	action := args[0]
	rest := args[1:]

	switch action {
	case "open":
		if len(rest) == 0 {
			return nil
		}
		return &browserRequest{Action: "open", URL: rest[0]}

	case "snapshot":
		req := &browserRequest{Action: "snapshot"}
		for _, a := range rest {
			if a == "--interactive" || a == "-i" {
				req.Interactive = true
			}
		}
		return req

	case "click":
		if len(rest) == 0 {
			return nil
		}
		return &browserRequest{Action: "click", Ref: normalizeRef(rest[0])}

	case "hover":
		if len(rest) == 0 {
			return nil
		}
		return &browserRequest{Action: "hover", Ref: normalizeRef(rest[0])}

	case "fill":
		if len(rest) < 2 {
			return nil
		}
		return &browserRequest{Action: "fill", Ref: normalizeRef(rest[0]), Text: strings.Join(rest[1:], " ")}

	case "type":
		if len(rest) < 2 {
			return nil
		}
		return &browserRequest{Action: "type", Ref: normalizeRef(rest[0]), Text: strings.Join(rest[1:], " ")}

	case "press":
		if len(rest) == 0 {
			return nil
		}
		return &browserRequest{Action: "press", Key: rest[0]}

	case "scroll":
		if len(rest) == 0 {
			return nil
		}
		req := &browserRequest{Action: "scroll", Direction: rest[0]}
		if len(rest) > 1 {
			var px int
			if _, err := fmt.Sscanf(rest[1], "%d", &px); err == nil {
				req.Pixels = &px
			}
		}
		return req

	case "eval":
		if len(rest) == 0 {
			return nil
		}
		return &browserRequest{Action: "eval", Script: strings.Join(rest, " ")}

	case "get":
		if len(rest) == 0 {
			return nil
		}
		req := &browserRequest{Action: "get", Attribute: rest[0]}
		if len(rest) > 1 {
			req.Ref = normalizeRef(rest[1])
		}
		return req

	case "screenshot":
		return &browserRequest{Action: "screenshot"}

	case "close":
		return &browserRequest{Action: "close"}

	case "back":
		return &browserRequest{Action: "back"}

	case "forward":
		return &browserRequest{Action: "forward"}

	case "refresh":
		return &browserRequest{Action: "refresh"}

	case "tabs":
		return &browserRequest{Action: "tab_list"}

	case "tab-new":
		req := &browserRequest{Action: "tab_new"}
		if len(rest) > 0 {
			req.URL = rest[0]
		}
		return req

	case "tab-select":
		if len(rest) == 0 {
			return nil
		}
		return &browserRequest{Action: "tab_select", TabID: rest[0]}

	case "tab-close":
		req := &browserRequest{Action: "tab_close"}
		if len(rest) > 0 {
			req.TabID = rest[0]
		}
		return req

	default:
		return nil
	}
}

// normalizeRef strips leading @ from ref strings (e.g. "@5" → "5").
func normalizeRef(ref string) string {
	return strings.TrimPrefix(ref, "@")
}

File: src/browser.ts
import { invoke } from "@pinixai/core";
import { mkdirSync, writeFileSync } from "node:fs";
import { basename, join } from "node:path";
import type { Config } from "./config";
import { dataRoot } from "./paths";
import { humanSize, pinixDataURLPrefix } from "./fs";

interface BrowserPageInfo {
  url?: string;
  title?: string;
}

interface BrowserNavigateResult {
  url: string;
  title: string;
}

interface BrowserEvaluateResult {
  result: unknown;
}

interface BrowserScreenshotResult {
  base64: string;
}

interface BrowserCookiesResult {
  cookies: Array<{
    name: string;
    value: string;
    domain: string;
    path: string;
  }>;
}

export function registerBrowserCommands(
  register: (name: string, description: string, handler: (args: string[], stdin: string) => Promise<string>) => void,
  _cfg: Config,
): void {
  register(
    "browser",
    `Control the user's browser.
  browser open <url>                  - open URL in current tab
  browser snapshot [--interactive]    - get page snapshot with refs
  browser click <ref|selector>        - click element
  browser fill <ref|selector> <text>  - clear and fill text
  browser type <ref|selector> <text>  - type text without clearing
  browser press <key>                 - press key on active element
  browser scroll <dir> [pixels]       - scroll: up/down/left/right
  browser eval <script>               - execute JavaScript
  browser get text|url|title [ref]    - get text or page info
  browser screenshot                  - take screenshot
  browser close                       - request tab close
  browser back|forward|refresh        - navigate history
  browser tabs                        - show current tab info
  browser tab-new [url]               - request new tab
  browser tab-select <id>             - unsupported on current provider
  browser tab-close [id]              - unsupported on current provider`,
    async (args) => browserCommand(args),
  );
}

async function browserCommand(args: string[]): Promise<string> {
  if (args.length === 0) {
    throw new Error("usage: browser <action> [args...]");
  }

  const action = args[0];
  const rest = args.slice(1);

  switch (action) {
    case "open":
      if (rest.length === 0) {
        throw new Error("usage: browser open <url>");
      }
      return formatPageInfo(await browserNavigate(rest[0]));
    case "snapshot": {
      const interactive = rest.includes("--interactive") || rest.includes("-i");
      const result = await browserEvaluate(buildSnapshotScript(interactive));
      return stringifyBrowserValue(result.result);
    }
    case "click":
      if (rest.length === 0) {
        throw new Error("usage: browser click <ref|selector>");
      }
      await browserClick(rest[0]);
      return "OK";
    case "fill":
      if (rest.length < 2) {
        throw new Error("usage: browser fill <ref|selector> <text>");
      }
      await browserFill(rest[0], rest.slice(1).join(" "));
      return "OK";
    case "type":
      if (rest.length < 2) {
        throw new Error("usage: browser type <ref|selector> <text>");
      }
      await browserType(rest[0], rest.slice(1).join(" "));
      return "OK";
    case "press":
      if (rest.length === 0) {
        throw new Error("usage: browser press <key>");
      }
      await browserPress(rest[0]);
      return "OK";
    case "scroll":
      if (rest.length === 0) {
        throw new Error("usage: browser scroll <dir> [pixels]");
      }
      await browserScroll(rest[0], rest[1] ? Number.parseInt(rest[1], 10) : 300);
      return "OK";
    case "eval":
      if (rest.length === 0) {
        throw new Error("usage: browser eval <script>");
      }
      return stringifyBrowserValue((await browserEvaluate(rest.join(" "))).result);
    case "get":
      if (rest.length === 0) {
        throw new Error("usage: browser get text|url|title [ref]");
      }
      return browserGet(rest[0], rest[1]);
    case "screenshot":
      return browserScreenshot();
    case "back":
      return formatPageInfo(await browserNavigateHistory("back"));
    case "forward":
      return formatPageInfo(await browserNavigateHistory("forward"));
    case "refresh":
      return formatPageInfo(await browserRefresh());
    case "close":
      await browserEvaluate("window.close(); true;");
      return "Close requested.";
    case "tabs":
      return formatPageInfo(await getCurrentPageInfo());
    case "tab-new":
      await browserEvaluate(`window.open(${JSON.stringify(rest[0] ?? "about:blank")}, '_blank'); true;`);
      return "New tab requested.";
    case "tab-select":
    case "tab-close":
      throw new Error(`${action} is not supported by the current browser provider`);
    default:
      throw new Error(`unknown browser action: ${action}`);
  }
}

async function browserNavigate(url: string): Promise<BrowserNavigateResult> {
  return await invoke("browser", "navigate", { url, waitUntil: "domcontentloaded" }) as BrowserNavigateResult;
}

async function browserClick(refOrSelector: string): Promise<void> {
  await invoke("browser", "click", { selector: selectorFromRef(refOrSelector) });
}

async function browserType(refOrSelector: string, text: string): Promise<void> {
  await invoke("browser", "type", { selector: selectorFromRef(refOrSelector), text });
}

async function browserFill(refOrSelector: string, text: string): Promise<void> {
  await browserEvaluate(`(() => {
    const el = document.querySelector(${JSON.stringify(selectorFromRef(refOrSelector))});
    if (!el) throw new Error(${JSON.stringify(`Element not found: ${refOrSelector}`)});
    if ('value' in el) {
      el.value = ${JSON.stringify(text)};
      el.dispatchEvent(new Event('input', { bubbles: true }));
      el.dispatchEvent(new Event('change', { bubbles: true }));
      return true;
    }
    if (el.isContentEditable) {
      el.textContent = ${JSON.stringify(text)};
      el.dispatchEvent(new Event('input', { bubbles: true }));
      return true;
    }
    throw new Error('Element is not editable');
  })()`);
}

async function browserPress(key: string): Promise<void> {
  await browserEvaluate(`(() => {
    const target = document.activeElement || document.body;
    const init = { key: ${JSON.stringify(key)}, bubbles: true };
    target.dispatchEvent(new KeyboardEvent('keydown', init));
    target.dispatchEvent(new KeyboardEvent('keypress', init));
    target.dispatchEvent(new KeyboardEvent('keyup', init));
    if (${JSON.stringify(key)} === 'Enter' && target instanceof HTMLElement && typeof target.click === 'function') {
      const form = target.closest('form');
      if (form) {
        form.dispatchEvent(new Event('submit', { bubbles: true, cancelable: true }));
      }
    }
    return true;
  })()`);
}

async function browserScroll(direction: string, pixels: number): Promise<void> {
  const normalized = Number.isFinite(pixels) && pixels > 0 ? pixels : 300;
  const delta = {
    up: { x: 0, y: -normalized },
    down: { x: 0, y: normalized },
    left: { x: -normalized, y: 0 },
    right: { x: normalized, y: 0 },
  }[direction];
  if (!delta) {
    throw new Error(`invalid scroll direction: ${direction}`);
  }
  await browserEvaluate(`window.scrollBy(${delta.x}, ${delta.y}); true;`);
}

async function browserGet(attribute: string, refOrSelector?: string): Promise<string> {
  switch (attribute) {
    case "url":
      return (await getCurrentPageInfo()).url || "";
    case "title":
      return (await getCurrentPageInfo()).title || "";
    case "text": {
      const selector = refOrSelector ? selectorFromRef(refOrSelector) : "body";
      const result = await browserEvaluate(`(() => {
        const el = document.querySelector(${JSON.stringify(selector)});
        if (!el) throw new Error(${JSON.stringify(`Element not found: ${refOrSelector ?? selector}`)});
        return (el.innerText || el.textContent || '').trim();
      })()`);
      return stringifyBrowserValue(result.result);
    }
    default:
      throw new Error(`unsupported browser get attribute: ${attribute}`);
  }
}

async function browserEvaluate(js: string): Promise<BrowserEvaluateResult> {
  return await invoke("browser", "evaluate", { js }) as BrowserEvaluateResult;
}

async function browserScreenshot(): Promise<string> {
  const screenshot = await invoke("browser", "screenshot", {}) as BrowserScreenshotResult;
  const bytes = Buffer.from(screenshot.base64, "base64");
  const filename = `screenshot-${Date.now()}.png`;
  const imageDir = dataRoot("images");
  mkdirSync(imageDir, { recursive: true });
  writeFileSync(join(imageDir, filename), bytes);

  const pageInfo = await getCurrentPageInfo();
  const parts = [];
  if (pageInfo.title) {
    parts.push(`Title: ${pageInfo.title}`);
  }
  if (pageInfo.url) {
    parts.push(`URL: ${pageInfo.url}`);
  }
  parts.push(`Screenshot: ${filename} (${humanSize(bytes.byteLength)})`);
  parts.push(`Render: ![screenshot](${pinixDataURLPrefix}${join("images", basename(filename))})`);
  return parts.join("\n");
}

async function browserNavigateHistory(direction: "back" | "forward"): Promise<BrowserPageInfo> {
  await browserEvaluate(`history.${direction}(); true;`);
  await delay(600);
  return getCurrentPageInfo();
}

async function browserRefresh(): Promise<BrowserPageInfo> {
  const current = await getCurrentPageInfo();
  if (current.url) {
    return browserNavigate(current.url);
  }
  await browserEvaluate("location.reload(); true;");
  await delay(600);
  return getCurrentPageInfo();
}

async function getCurrentPageInfo(): Promise<BrowserPageInfo> {
  const result = await browserEvaluate("JSON.stringify({ url: location.href, title: document.title })");
  if (typeof result.result === "string") {
    try {
      return JSON.parse(result.result) as BrowserPageInfo;
    } catch {
      return {};
    }
  }
  return (result.result as BrowserPageInfo) ?? {};
}

function formatPageInfo(info: BrowserPageInfo): string {
  const parts: string[] = [];
  if (info.title) {
    parts.push(`Title: ${info.title}`);
  }
  if (info.url) {
    parts.push(`URL: ${info.url}`);
  }
  return parts.length > 0 ? parts.join("\n") : "OK";
}

function selectorFromRef(refOrSelector: string): string {
  const trimmed = refOrSelector.trim();
  if (/^@?\d+$/.test(trimmed)) {
    return `[data-pinix-ref="${trimmed.replace(/^@/, "")}"]`;
  }
  return trimmed;
}

function stringifyBrowserValue(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }
  return JSON.stringify(value, null, 2);
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function buildSnapshotScript(interactiveOnly: boolean): string {
  return `(() => {
    const isVisible = (el) => {
      const rect = el.getBoundingClientRect();
      const style = window.getComputedStyle(el);
      return rect.width > 0 && rect.height > 0 && style.visibility !== 'hidden' && style.display !== 'none';
    };

    document.querySelectorAll('[data-pinix-ref]').forEach((el) => el.removeAttribute('data-pinix-ref'));

    const selector = ${JSON.stringify(interactiveOnly
      ? 'a,button,input,textarea,select,summary,[role="button"],[contenteditable="true"],[onclick]'
      : 'a,button,input,textarea,select,summary,[role="button"],[contenteditable="true"],[onclick],h1,h2,h3,p,li,article,section,main,nav')};
    const nodes = Array.from(document.querySelectorAll(selector));
    const lines = [
      'Title: ' + document.title,
      'URL: ' + location.href,
      ''
    ];
    let index = 1;
    for (const node of nodes) {
      if (!isVisible(node)) continue;
      node.setAttribute('data-pinix-ref', String(index));
      const tag = node.tagName.toLowerCase();
      const text = (node.innerText || node.textContent || '').replace(/\\s+/g, ' ').trim();
      const href = node instanceof HTMLAnchorElement ? node.href : '';
      const placeholder = 'placeholder' in node && typeof node.placeholder === 'string' ? node.placeholder : '';
      const label = [text, placeholder, href].filter(Boolean).join(' | ');
      lines.push('@' + index + ' <' + tag + '> ' + (label || '(empty)'));
      index += 1;
    }
    return lines.join('\\n');
  })()`;
}
